### PR TITLE
ISOM-852 feat(logger): add context for logs

### PR DESCRIPTION
**Pull request by seaerchin**
**https://linear.app/ogp/issue/ISOM-852/add-structured-logging-context**
## Problem
See [here](https://www.notion.so/opengov/Structured-logging-4072eeeeded3422bab38497b738096c4?pvs=4)

## Solution
1. create a basic `Logger` class. The base logger has only 1 method exposed (enforced via ts) so that consumers are forced to call `.child` and inject a module
2. stack traces are inserted via a v8 stack trace (with reference to `stack-trace` library). this provides useful info (up to a depth of 3) such as file, function etc. 
    - for `logger.error` calls, we instead use the error's stack trace (as the stack would've been popped already) if it is provided.
3. all meta information is injected via a `isomer` key so that there's no top level confusion. (maybe `module` should've been there too? idk)

### CODE REVIEW

**Review by timotheeg — COMMENTED** [review]
Nice work Chin! 

I'm not entirely convinced we needed to have that level of (costly?) beautification for stack traces 🤔 . Vanilla `pino` already reports errors with a "standard" structure, and traces therein as newline-separated strings with tabs. That might have been good enough operationally, and much cheaper to produce.

e.g.
```
pinoLogger.info(new Error('hello'))
```
gives:
```
{
  "level": 30,
  "time": 1711002385041,
  "pid": 62798,
  "hostname": "Timothee-Groleaus-MacBook-Pro-14-inch-2021-",
  "err": {
    "type": "Error",
    "message": "hello",
    "stack": "Error: hello\n    at REPL11:1:8\n    at Script.runInThisContext (node:vm:123:12)\n    at REPLServer.defaultEval (node:repl:569:29)\n    at bound (node:domain:433:15)\n    at REPLServer.runBound [as eval] (node:domain:444:12)\n    at REPLServer.onLine (node:repl:899:10)\n    at REPLServer.emit (node:events:526:35)\n    at REPLServer.emit (node:domain:489:12)\n    at [_onLine] [as _onLine] (node:internal/readline/interface:422:12)\n    at [_line] [as _line] (node:internal/readline/interface:893:18)"
  },
  "msg": "hello"
}
```

Also, a custom stack trace format might not work well with source code linkage in the future (I'm not sure exactly about that though).

Still, I say we can try with your approach first, and iterate as needed. Hopefully we don't log error objects too often, so the stack trace transformation cost should be negligible, but still, possibly the beautification wasn't needed in the first place 🤔


On the PR itself, might be helpful to show typical invocation and outputs so reviewers get a quick sense of what we'll be getting exactly.

File: src/logger/logger.ts [hunk 1]
```diff
@@ -1,12 +1,100 @@
...
+
+  public error = (
+    message: string,
+    meta: WarnMeta & ErrorMeta & Record<string, unknown>
+  ): void => {
+    const stackFrames = meta.error instanceof Error ? parse(meta.error) : get()
+    const trace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+    this._logger.error(message, {
+      isomer: {
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:51]: i'm not too opinionated wrt `isomer` key tbh, it's just that in some logs i've observed, the log metadata is grouped by the origin (`aws/dd` here) 

wrt `error/err` it's there already, as part of `meta` - `type meta = WarnMeta & ErrorMeta & Record<string, unknown>`

and `ErrorMeta` is `{ error: unknown }`. i chose `unknown` over `error` because in `try/catch` loops, the `error` is only allowed to be of type `any | unknown`. so rather let devs iterate fast here and pass in the `error` directly then `logger` does some extra work. 

tl;dr: we accept error object, the `trace` here is taken from the `stack` property of `error`

Screenshot 2024-03-21 at 4 13 44 PM

File: src/logger/logger.ts [hunk 2]
```diff
@@ -1,12 +1,100 @@
...
+
+interface WarnMeta {
+  params: Record<string, unknown>
+}
+
+interface ErrorMeta {
+  error: unknown
+}
+
+const MAX_STACK_DEPTH = 3
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:14]: sure! wrt errors passed by callers - the raw error itself is preserved, this only comes into play if caller doesn't pass an `error` (then we have to extract stack trace). 

wdyt about having it at 5? (same as pino recursive depth)

File: src/logger/logger.ts [hunk 3]
```diff
@@ -1,12 +1,100 @@
...
+
+export class Logger {
+  _logger: PinoLogger
+
+  constructor(logger: PinoLogger) {
+    this._logger = logger
+  }
+
+  public child = (
+    bindings: Record<string, unknown> & { module: string }
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:37]: sorry not too clear what you mean here! but `bindings` are objects that `pino` injects automatically to **every** log made by the logger. 

see [here](https://getpino.io/#/docs/api?id=loggerbindings) for docs wrt bindings; pino [preserves duplicate keys](https://getpino.io/#/docs/child-loggers?id=duplicate-keys-caveat) so i think the main worry is just confusion (if there are duplicate keys)

File: src/logger/logger.ts [hunk 4]
```diff
@@ -1,12 +1,100 @@
...
+    message: string,
+    meta: WarnMeta & ErrorMeta & Record<string, unknown>
+  ): void => {
+    const stackFrames = meta.error instanceof Error ? parse(meta.error) : get()
+    const trace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+    this._logger.error(message, {
+      isomer: {
+        meta,
+        trace,
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:53]: sounds good! will call it `stackTrace`

File: src/logger/logger.ts [hunk 5]
```diff
@@ -1,12 +1,100 @@
...
+export interface TraceMeta {
+  file: string
+  line: number
+  func: string
+}
+
+export const getTraceMeta = (frames: StackFrame[]): TraceMeta[] =>
+  frames.map((frame) => ({
+    file: frame.getFileName(),
+    line: frame.getLineNumber(),
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:25]: hmm lemme check

File: src/logger/logger.ts [hunk 6]
```diff
@@ -1,12 +1,100 @@
...
+    const child = this._logger.child(bindings)
+    return new Logger(child)
+  }
+
+  public error = (
+    message: string,
+    meta: WarnMeta & ErrorMeta & Record<string, unknown>
+  ): void => {
+    const stackFrames = meta.error instanceof Error ? parse(meta.error) : get()
+    const trace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:48]: the `error` itself is still retained, so idt this would matter!

File: src/logger/logger.ts [hunk 7]
```diff
@@ -1,12 +1,100 @@
...
+
+  public child = (
+    bindings: Record<string, unknown> & { module: string }
+  ): Logger => {
+    const child = this._logger.child(bindings)
+    return new Logger(child)
+  }
+
+  public error = (
+    message: string,
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:44]: we could! do take note that most errors wouldn't be known at compile time to be of type `Error` (ie, devs will pass in `error: unknown` and we'd do the validation ourselves). 

this has some amount of repetition tbh (`message` could be equivalent to `error.message`) but yea i'm ok with this for now. wdyt?

File: src/logger/logger.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:25]: i think it maps from the error itself so it does preserve source mapping!

File: src/logger/logger.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:14]: updated to 5

File: src/logger/logger.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:53]: updated

File: src/logger/logger.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (timotheeg) [reply src/logger/logger.ts:25]: Nice!

File: src/logger/logger.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (timotheeg) [reply src/logger/logger.ts:37]: Thanks for that! I was indeed wondering what happens when there is a conflicts between the bindings and data that gets passed at call time.

Outputting duplicate keys is an interesting take, as human we can see the content so nothing is lost, but a json parser would only keep the last value, so the duplicate keys are not very useful in practice. We should take note to not have duplicate keys.

File: src/logger/logger.ts [hunk 8]
```diff
@@ -1,12 +1,100 @@
...
+
+  public info = (message: string, meta?: Record<string, unknown>): void => {
+    if (_.isEmpty(meta)) {
+      this._logger.info(message)
+    } else {
+      this._logger.info(message, {
+        isomer: {
+          meta,
+        },
+      })
```
Comment (timotheeg) [inline src/logger/logger.ts:66]: ⚠️ Does this call signature actually work? [Documentation](https://github.com/pinojs/pino/blob/master/docs/api.md#info) says it should be object first then message.

Quick test to confirm:

```javascript
pino = require('pino')
logger =   pino({
    errorKey: "error",
    messageKey: "message",
    formatters: {
      level: (label) => ({ level: label.toUpperCase() }),
    },
  })

logger.info('hello', {meta: {foo: 1}})
```
gives
```
{"level":"INFO","time":1711066498870,"pid":69756,"hostname":"Timothee-Groleaus-MacBook-Pro-14-inch-2021-","message":"hello"}
```

On the other hand, this works as expected:
```javascript
logger.info({meta: {foo: 1}}, 'hello')
```
gives
```
{"level":"INFO","time":1711066992960,"pid":72744,"hostname":"Timothee-Groleaus-MacBook-Pro-14-inch-2021-","meta":{"foo":1},"message":"hello"}
```

File: src/logger/logger.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (timotheeg) [reply src/logger/logger.ts:51]: > it's just that in some logs i've observed, the log metadata is grouped by the origin (aws/dd here)

Ah got it, I understand your intent now! It makes sense for providers andplugins to dump their values in namespaces, since that reduces chances of collision with what the app logs. In our case, since we are the top-level app, there's no need for an additional namespace, we could argue that even the `meta` envelop could be considered unnecessary, but I think it's cheap enough to keep 🤔 


> wrt error/err it's there already, as part of meta - type meta = WarnMeta & ErrorMeta & Record<string, unknown>

But that won't work (in the sense that it won't output any useful to the logs).

Error objects are not iterable/serializable and need custom serialization to work. With the current code, you'd get an empty object in the log:

Example:
```javascript
logger.error({meta: {error: new Error('foo')}})
```
```
{
  "level":"ERROR",
  "time":1711067699209,
  "pid":76651,
  "hostname":"Timothee-Groleaus-MacBook-Pro-14-inch-2021-",
  "meta":{
    "error":{}
  }
}
```

[the remark this answers is not in the document]
Comment (timotheeg) [reply src/logger/logger.ts:51]: Oh, on the namespace thing, pino does have a [built-in functionality](https://github.com/pinojs/pino/blob/master/docs/api.md#nestedkey-string) for that btw

**Review by timotheeg — CHANGES_REQUESTED** [review]
Hey, so I dived into the pino docs a bit more, and so I have a few higher level recommendation for the logger system:

1. as a principle, since we've chosen pino, let's use its convention where possible instead of shoehorning our own into it. For example, pino can run serializers on top level properties of the mergingObject, so special objects we are interested in processing (e.g. errors) should be top level of the mergingObject, and not nested into a meta namespace
2. Let's not create a wrapper class for Logger. Restore exposing a root pino logger from logger.ts. With that, the app will have the full pino instance API at its disposal if needed (e.g. printf interpolations if we ever want/need to use it).
3. If we still want to beautify the stack traces, use the pino's system of [serializers](https://github.com/pinojs/pino/blob/master/docs/api.md#serializers-object) and make a custom serializer that parses the trace to outputs it in a clean way. We need to make sure our custom Error serializer outputs the error message, and optional fields if present (e.g. `code`, `status`, or `cause`). These fields are not standard error fields, but many libs adds them to their error objects.

On 2), I think from your current code, one custom behaviour is to generate local error objects for the `warn` and `error` methods to get their call traces. I believe that can be achieved with a [pino hook](https://github.com/pinojs/pino/blob/master/docs/api.md#hooks-object) (to be confirmed / tested)

Like I mentioned in my previous comment, I'm not sure beautifying error traces was the best use of time 😅 , but since you've already done it, porting the work into a pino serializer should not take too long.

File: src/logger/logger.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:51]: as discussed, removed the namespace (also thanks for the link!) and opted to stick with a top level `meta` key

File: src/logger/logger.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/logger/logger.ts:44]: changed API to take `unknown` and work with the given `error` param from there!

File: src/logger/logger.ts (hunk 8 above)
Comment (seaerchin) [reply src/logger/logger.ts:66]: gud ketch, fixed!

**Comment by seaerchin** [comment]
> 1. as a principle, since we've chosen pino, let's use its convention where possible instead of shoehorning our own into it. For example, pino can run serializers on top level properties of the mergingObject, so special objects we are interested in processing (e.g. errors) should be top level of the mergingObject, and not nested into a meta namespace

this has been fixed!

> 2. Let's not create a wrapper class for Logger. Restore exposing a root pino logger from logger.ts. With that, the app will have the full pino instance API at its disposal if needed (e.g. printf interpolations if we ever want/need to use it).

as discussed offline, we'll wrap it first due to hooks requiring some casting due to `pino` being in js, which creates confusion. callers still have access via `logger._logger`

> 3. If we still want to beautify the stack traces, use the pino's system of [serializers](https://github.com/pinojs/pino/blob/master/docs/api.md#serializers-object) and make a custom serializer that parses the trace to outputs it in a clean way. We need to make sure our custom Error serializer outputs the error message, and optional fields if present (e.g. `code`, `status`, or `cause`). These fields are not standard error fields, but many libs adds them to their error objects.
> 
> On 2), I think from your current code, one custom behaviour is to generate local error objects for the `warn` and `error` methods to get their call traces. I believe that can be achieved with a [pino hook](https://github.com/pinojs/pino/blob/master/docs/api.md#hooks-object) (to be confirmed / tested)

see above!

**Comment by kishore03109** [comment]
>as discussed offline, we'll wrap it first due to hooks requiring some casting due to pino being in js, which creates confusion. callers still have access via logger._logger

@seaerchin why are we not considering just adding a d.ts ah?

File: src/logger/logger.ts [hunk 9]
```diff
@@ -1,12 +1,139 @@
...
+    } else if (messageOrError instanceof Error) {
+      const stackFrames = parse(messageOrError)
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+      this._logger.error({
+        error: messageOrError,
+        stackTrace,
+      })
+    } else if (hasErrorProp(messageOrError)) {
+      const stackFrames = parse(messageOrError?.error)
```
Comment (timotheeg) [inline src/logger/logger.ts:71]: ⚠️ `messageOrError.error` is not guaranteed to be an instance of Error wit the current check Would `parse()` barf at it?

File: src/logger/logger.ts [hunk 10]
```diff
@@ -1,12 +1,139 @@
...
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+      this._logger.error({
+        error: messageOrError.error,
+        meta: _.omit(messageOrError, "error"),
+        stackTrace,
+      })
+    } else {
+      const stackFrames = get()
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
```
Comment (timotheeg) [inline src/logger/logger.ts:81]: Bit of repetition of the `slice()` calls, we could shift that into `getTraceMeta()`

File: src/logger/logger.ts [hunk 11]
```diff
@@ -1,12 +1,139 @@
...
+    } else {
+      const meta = _.omit(messageOrObject, "message")
+      this._logger.info({ meta }, messageOrObject.message)
+    }
+  }
+
+  public warn = (
+    messageOrObject:
+      | string
+      | (WarnMeta & Record<string, unknown> & { message: string })
```
Comment (timotheeg) [inline src/logger/logger.ts:104]: I think the typings can be improved a bit 🤔

File: src/logger/logger.ts [hunk 12]
```diff
@@ -1,12 +1,139 @@
-import pino from "pino"
+import _ from "lodash"
+import pino, { Logger as PinoLogger } from "pino"
+
+import { StackFrame, get, parse } from "./stack-trace"
+
+interface WarnMeta {
+  params: Record<string, unknown>
```
Comment (timotheeg) [inline src/logger/logger.ts:7]: Why do we need params in WarnMeta?

File: src/logger/logger.ts [hunk 13]
```diff
@@ -1,12 +1,139 @@
...
+      this._logger.error({
+        error: messageOrError,
+        stackTrace,
+      })
+    } else if (hasErrorProp(messageOrError)) {
+      const stackFrames = parse(messageOrError?.error)
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+      this._logger.error({
+        error: messageOrError.error,
```
Comment (timotheeg) [inline src/logger/logger.ts:75]: Just noting: when error is an `Error`, pino will still use its standard formatter, and so the log line will have some repetition on the stack trace.

**Comment by timotheeg** [comment]
@seaerchin I pushed some changes. Please check. 

Just a coupe of quick notes:
* The release today is quite heavy (potential cpu cost with the trace PR), and this logger PR potentially adds costs too with stack frame processing for all `warn()` and `error()` log calls. Because of that, we'll push it separately to be able to monitor potential issues better.
* For the cases where error is `unknown` (when we just log as-is whatever it is that was caught), we should eventually change to instantiate our own errors, and pass the source as [cause](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause). We can use pino's [`errWithCause` serialiser](https://github.com/pinojs/pino-std-serializers?tab=readme-ov-file#exportserrwithcauseerror) to get all the data outputted nicely for us.

File: src/logger/logger.ts (hunk 12 above)
Comment (seaerchin) [reply src/logger/logger.ts]: my assumption is that when people call `warn`, it's dangerous behaviour that should be looked at still. hence, we want to capture as much context as possible (such as `params` of the calling function) to aid in debugging

File: src/logger/logger.ts (hunk 9 above)
Comment (seaerchin) [reply src/logger/logger.ts]: good catch, going to make the check more stringent

Comment (seaerchin) [reply src/logger/logger.ts]: oh saw that yuo did it alr, marking as resolved

File: src/logger/logger.ts (hunk 11 above)
Comment (seaerchin) [reply src/logger/logger.ts]: thanks for doing this, got lazy ._.

**Review by timotheeg — APPROVED** [review]
